File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/routing/StandardHandlers.java
package uk.nhs.nhsx.core.routing;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import org.apache.http.entity.ContentType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import uk.nhs.nhsx.core.Environment;
import uk.nhs.nhsx.core.HttpResponses;
import uk.nhs.nhsx.core.auth.Authenticator;
import uk.nhs.nhsx.core.auth.ResponseSigner;
import uk.nhs.nhsx.core.exceptions.ApiResponseException;

import java.util.Base64;
import java.util.Optional;
import java.util.function.Consumer;

public class StandardHandlers {

    private static final Logger logger = LogManager.getLogger(StandardHandlers.class);
    
    private static final Environment.EnvironmentKey<String> MAINTENANCE_MODE = Environment.EnvironmentKey.string("MAINTENANCE_MODE");
    private static final Environment.EnvironmentKey<String> CUSTOM_OAI = Environment.EnvironmentKey.string("custom_oai");
    

    public static Routing.Handler withoutSignedResponses(Environment environment, Authenticator authenticator, Routing.Handler delegate) {
        return defaultStack(environment, authenticator, catchExceptions(delegate));
    }

    public static Routing.Handler withSignedResponses(Environment environment, Authenticator authenticator, ResponseSigner signer, Routing.Handler delegate) {
        return defaultStack(environment, authenticator, signedBy(signer, catchExceptions(delegate)));
    }

    private static Routing.Handler defaultStack(Environment environment, Authenticator authenticator, Routing.Handler handler) {
        return loggingIncomingRequests(
                filteringWhileMaintenanceModeEnabled(environment,
                requiringAuthorizationHeader(
                    requiringCustomAccessIdentity(environment,
                        authorisedBy(authenticator,
                            handler
                        )
                    )
                )
            )
        );
    }

    public static Routing.Handler loggingIncomingRequests(Routing.Handler delegate) {
        return r -> {
            String keyName =  apiKeyNameFrom(r.getHeaders().get("authorization")).orElse("none");
            String requestId = Optional.ofNullable(r.getHeaders().get("Request-Id")).orElse("none");
            logger.info("Received http request: method={}, path={},requestId={},apiKeyName={}", r.getHttpMethod(), r.getPath(),requestId,keyName);
            return delegate.handle(r);
        };
    }

    public static Routing.Handler filteringWhileMaintenanceModeEnabled(Environment environment, Routing.Handler delegate) {
        if (Boolean.parseBoolean(environment.access.required(MAINTENANCE_MODE))) {
            return r -> HttpResponses.serviceUnavailable();
        }
        else {
            return delegate;
        }
    }

    public static Routing.Handler requiringAuthorizationHeader(Routing.Handler delegate) {
        return r -> Optional.ofNullable(r.getHeaders().get("authorization")).map(x -> delegate.handle(r)).orElse(HttpResponses.forbidden());
    }

    public static Routing.Handler requiringCustomAccessIdentity(Environment environment, Routing.Handler delegate) {
        Optional<String> maybeRequiredOai = environment.access.optional(CUSTOM_OAI);

        return maybeRequiredOai.map(
            requiredOai -> (Routing.Handler) request -> {
                if (requiredOai.equals(request.getHeaders().get("x-custom-oai"))) {
                    logger.debug("Access to API Gateway via CloudFront facade (with x-custom-oai): method={}, path={}", request.getHttpMethod(), request.getPath());
                    return delegate.handle(request);
                } else {
                    logger.warn("Direct access to API Gateway (without x-custom-oai): method={}, path={}", request.getHttpMethod(), request.getPath());
                    return HttpResponses.forbidden();
                }
            }
        ).orElse(
            request -> {
                logger.error("API protection not configured (no expected x-custom-oai): method={}, path={}", request.getHttpMethod(), request.getPath());
                return delegate.handle(request);
            }
        );
    }

    public static Routing.Handler authorisedBy(Authenticator authenticator, Routing.Handler delegate) {
        return r -> Optional.ofNullable(r.getHeaders().get("authorization"))
            .filter(authenticator::isAuthenticated)
            .map(a -> delegate.handle(r))
            .orElse(HttpResponses.forbidden());
    }

    public static Routing.Handler signedBy(ResponseSigner signer, Routing.Handler delegate) {
        return request -> {
            APIGatewayProxyResponseEvent response = delegate.handle(request);
            signer.sign(request, response);
            return response;
        };
    }

    public static Routing.Handler responseFilter(Routing.Handler delegate, Consumer<APIGatewayProxyResponseEvent> fiddler) {
        return (r) -> {
            APIGatewayProxyResponseEvent response = delegate.handle(r);
            fiddler.accept(response);
            return response;
        };
    }

    public static Routing.Handler expectingContentType(ContentType contentType, Routing.Handler handler) {
        return r -> {
            Optional<String> given = Optional.ofNullable(r.getHeaders().get("Content-Type"));
            if (given.isPresent()) {
                return given
                    .filter(c -> contentType.getMimeType().equals(ContentType.parse(c).getMimeType()))
                    .map(c -> handler.handle(r))
                    .orElse(HttpResponses.unprocessableEntity());
            } else {
                return HttpResponses.badRequest();
            }
        };
    }

    public static Routing.Handler catchExceptions(Routing.Handler delegate) {
        return (r) -> {
            try {
                return delegate.handle(r);
            } catch (ApiResponseException e) {
                logger.info("Failed to handle request: error={}, returned status code={}", e.getMessage(), e.statusCode.code);
                return HttpResponses.withStatusCodeAndBody(e.statusCode, e.getMessage());
            } catch (Exception e) {
                logger.error("Uncaught exception {}", e.getMessage(), e);
                return HttpResponses.internalServerError();
            }
        };
    }

    public static Optional<String> apiKeyNameFrom(String authorizationHeader){
        return Optional.ofNullable(authorizationHeader)
            .filter(it -> it.startsWith("Bearer "))
            .map(it -> it.replaceFirst("Bearer ", ""))
            .flatMap(it -> {
                try {
                    byte[] decode = Base64.getDecoder().decode(it);
                    return Optional.of(new String(decode));
                } catch (IllegalArgumentException e) {
                    return Optional.empty();
                }
            })
            .flatMap(it -> {
                int colonPos = it.indexOf(":");
                if (colonPos == -1) return Optional.empty();
                return Optional.of(it.substring(0, colonPos));
            });
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/routing/StandardHandlers.kt
package uk.nhs.nhsx.core.routing

import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent
import org.apache.http.entity.ContentType
import uk.nhs.nhsx.core.Environment
import uk.nhs.nhsx.core.Environment.EnvironmentKey
import uk.nhs.nhsx.core.HttpResponses
import uk.nhs.nhsx.core.SystemClock
import uk.nhs.nhsx.core.auth.ApiKeyExtractor
import uk.nhs.nhsx.core.auth.Authenticator
import uk.nhs.nhsx.core.auth.ResponseSigner
import uk.nhs.nhsx.core.events.ApiHandleFailed
import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.core.events.ExceptionThrown
import uk.nhs.nhsx.core.events.IncomingHttpRequest
import uk.nhs.nhsx.core.events.OAINotSet
import uk.nhs.nhsx.core.events.RequestRejected
import uk.nhs.nhsx.core.exceptions.ApiResponseException
import uk.nhs.nhsx.core.headers.MobileAppVersion
import uk.nhs.nhsx.core.headers.UserAgent
import java.time.Duration
import java.time.Instant
import java.util.Optional
import java.util.function.Supplier

object StandardHandlers {
    private val MAINTENANCE_MODE = EnvironmentKey.string("MAINTENANCE_MODE")
    private val CUSTOM_OAI = EnvironmentKey.string("custom_oai")

    @JvmStatic
    fun withoutSignedResponses(
        events: Events,
        environment: Environment,
        delegate: ApiGatewayHandler
    ): ApiGatewayHandler =
        defaultStack(events, environment, catchExceptions(events, delegate))

    @JvmStatic
    fun withSignedResponses(
        events: Events,
        environment: Environment,
        signer: ResponseSigner,
        delegate: ApiGatewayHandler
    ): ApiGatewayHandler = defaultStack(events, environment, signedBy(signer, catchExceptions(events, delegate)))

    private fun defaultStack(events: Events, environment: Environment, handler: ApiGatewayHandler): ApiGatewayHandler =
        loggingIncomingRequests(
            events,
            filteringWhileMaintenanceModeEnabled(
                events,
                environment,
                requiringAuthorizationHeader(
                    requiringCustomAccessIdentity(events, environment, handler)
                )
            ),
            SystemClock.CLOCK
        )

    fun loggingIncomingRequests(
        events: Events,
        delegate: ApiGatewayHandler,
        clock: Supplier<Instant>
    ): ApiGatewayHandler =
        ApiGatewayHandler { r: APIGatewayProxyRequestEvent, context ->
            val keyName = ApiKeyExtractor(r.headers["authorization"])?.keyName ?: "none"
            val requestId = Optional.ofNullable(r.headers["Request-Id"]).orElse("none")
            val userAgent = userAgentFrom(r)

            val start = clock.get()
            var statusCode = 500
            try {
                delegate.invoke(r, context).also { statusCode = it.statusCode }
            } finally {
                val latency = Duration.between(start, clock.get())
                events(
                    javaClass,
                    IncomingHttpRequest(
                        r.path,
                        r.httpMethod,
                        statusCode,
                        latency.toMillis(),
                        UserAgent.of(userAgent),
                        requestId,
                        keyName,
                        "Received http request: method=${r.httpMethod},path=${r.path},requestId=${requestId},apiKeyName=${keyName},userAgent=${userAgent},status=${statusCode},latency=${latency}"
                    )
                )
            }
        }

    private fun userAgentFrom(r: APIGatewayProxyRequestEvent): String =
        Optional.ofNullable(r.headers["User-Agent"]).orElse("none")

    @JvmStatic
    fun mobileAppVersionFrom(r: APIGatewayProxyRequestEvent): MobileAppVersion =
        UserAgent.of(userAgentFrom(r)).appVersion

    fun filteringWhileMaintenanceModeEnabled(
        events: Events,
        environment: Environment,
        delegate: ApiGatewayHandler
    ): ApiGatewayHandler =
        when {
            environment.access.required(MAINTENANCE_MODE).toLowerCase().toBoolean() ->
                ApiGatewayHandler { _, _ ->
                    events(StandardHandlers::class.java, RequestRejected("MAINTENANCE_MODE"))
                    HttpResponses.serviceUnavailable()
                }
            else -> delegate
        }


    fun requiringAuthorizationHeader(delegate: ApiGatewayHandler): ApiGatewayHandler =
        ApiGatewayHandler { r: APIGatewayProxyRequestEvent, context ->
            Optional.ofNullable(r.headers["authorization"]).map { delegate.invoke(r, context) }
                .orElse(HttpResponses.forbidden())
        }

    fun requiringCustomAccessIdentity(
        events: Events,
        environment: Environment,
        delegate: ApiGatewayHandler
    ): ApiGatewayHandler =
        environment.access.required(CUSTOM_OAI).let { requiredOai: String ->
            ApiGatewayHandler { request: APIGatewayProxyRequestEvent, context ->
                if (requiredOai == request.headers["x-custom-oai"]) {
                    delegate.invoke(request, context)
                } else {
                    events.emit(javaClass, OAINotSet(request.httpMethod, request.path))
                    HttpResponses.forbidden()
                }
            }
        }

    @JvmStatic
    fun authorisedBy(authenticator: Authenticator, routingHandler: Routing.RoutingHandler): Routing.RoutingHandler =
        object : DelegatingRoutingHandler(routingHandler) {
            override fun invoke(request: APIGatewayProxyRequestEvent, context: Context): APIGatewayProxyResponseEvent =
                Optional.ofNullable(request.headers["authorization"])
                    .filter { authorizationHeader -> authenticator.isAuthenticated(authorizationHeader) }
                    .map { delegate.invoke(request, context) }
                    .orElse(HttpResponses.forbidden())
        }

    fun signedBy(signer: ResponseSigner, delegate: ApiGatewayHandler): ApiGatewayHandler =
        ApiGatewayHandler { request: APIGatewayProxyRequestEvent, context ->
            val response = delegate.invoke(request, context)
            if (response.statusCode != 403) {
                signer.sign(request, response)
            }
            response
        }

    fun expectingContentType(contentType: ContentType, handler: ApiGatewayHandler): ApiGatewayHandler =
        ApiGatewayHandler { r: APIGatewayProxyRequestEvent, context ->
            val given = Optional.ofNullable(r.headers["Content-Type"])
            if (given.isPresent) {
                given
                    .filter { c -> contentType.mimeType == ContentType.parse(c).mimeType }
                    .map { handler(r, context) }
                    .orElse(HttpResponses.unprocessableEntity())
            } else {
                HttpResponses.badRequest()
            }
        }

    fun catchExceptions(events: Events, delegate: ApiGatewayHandler): ApiGatewayHandler =
        ApiGatewayHandler { r: APIGatewayProxyRequestEvent, context ->
            try {
                delegate.invoke(r, context)
            } catch (e: ApiResponseException) {
                events.emit(javaClass, ApiHandleFailed(e.statusCode.code, e.message))
                HttpResponses.withStatusCodeAndBody(e.statusCode, e.message)
            } catch (e: Exception) {
                events.emit(javaClass, ExceptionThrown(e))
                HttpResponses.internalServerError()
            }
        }
}
